proposals/gc/Overview.md

```diff
@@ -1 +1,728 @@
+* Independent from linear memory.
+* Pay as you go; in particular, no effect on code not using GC.
+* Don't introduce dependencies on GC for other features (e.g., using resources through tables).
+* Avoid generics or other complex type structure _if possible_.
+* Extend the design iteratively, ship a minimal set of functionality fast.
+
+
+### Requirements
+
+* Allocation of structures that are garbage collected.
+* Allocation of unstructured byte arrays that are garbage collected.
+* Handles to heap values from the embedder, garbage collected.
+* Manipulating references to these, as value types.
+* Forming unions of different types, as value types? (future extension?)
+* Defining, allocating, and indexing structures as extensions to imported types? (future extension)
+* Exceptions (separate proposal)
+* Direct support for strings? (separate proposal)
+* Safe interaction with threads (sharing, atomic access)
+
+
+### Efficiency Considerations
+
+GC support should maintain Wasm's efficiency properties as much as possible, namely:
+
+* all operations are very cheap, ideally constant time,
+* structures are contiguous, dense chunks of memory,
+* accessing fields are single-indirection loads and stores,
+* allocation is fast,
+* no implicit boxing operations (i.e. no implicit allocation on the heap),
+* primitive values should not need to be boxed to be stored in managed data structures,
+* allows ahead-of-time compilation and code caching.
+
+
+### Evaluation
+
+Should attempt to implement 2-3 exemplary languages:
+
+* an object-oriented language (e.g., a subset of Java, with classes, inheritence, interfaces),
+* a typed functional language (e.g., a subset of ML, with closures, polymorphism, variant types)
+* an untyped language (e.g., a subset of Scheme or Python or something else)
+
+
+## Use Cases
+
+### Tuples and Arrays
+
+* Want to represent first-class tuples/records/structs with static indexing.
+* Want to represent arrays with dynamic indexing.
+* Possibly want to create arrays with both fixed or dynamic size.
+
+Example (fictional language):
+```
+type tup = (int, int, bool)
+type vec3d = float[3]
+type buf = {pos : int, buf : char[]}
+```
+
+Needs:
+
+* user-defined structures and arrays as heap objects,
+* references to those as first-class values.
+
+
+### Objects and Method Tables
+
+* Want to represent instances as structures, whose first field is the method table.
+* Want to represent method tables themselves as structures, whose fields are function pointers.
+* Subtyping is relevant, both on instance types and method table types.
+
+Example (Java-ish):
+```
+class C {
+  int a;
+  void f(int i);
+  int g();
+}
+class D extends C {
+  double b;
+  override int g();
+  int h();
+}
+```
+
+```
+(type $f-sig (func (param (ref $C)) (param i32)))   ;; first param is `this`
+(type $g-sig (func (param (ref $C)) (result i32)))
+(type $h-sig (func (param (ref $D)) (result i32)))
+
+(type $C (struct (ref $C-vt) (mut i32))
```

Comment:
Ah, I just realized that this "vtable" structure is not necessarily a structure of function pointers. It can contain all these fields as-is.